libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt

```diff
@@ -13,70 +16,69 @@ import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
  *
  * On back pressed will update the navigation state.
  *
  * (FreeDrive) <- (DestinationPreview) <- (RoutePreview) <- (ActiveNavigation)
  *             <- (Arrival)
  */
 @ExperimentalPreviewMapboxNavigationAPI
-internal class OnKeyListenerComponent(
+internal class DefaultOnBackPressedCallback private constructor(
     private val store: Store,
-    private val view: View,
-    private val delegateOnKeyListener: View.OnKeyListener? = null
-) : UIComponent() {
+) : OnBackPressedCallback(false) {
 
-    override fun onAttached(mapboxNavigation: MapboxNavigation) {
-        super.onAttached(mapboxNavigation)
-
-        view.isFocusableInTouchMode = true
-        view.requestFocus()
-        view.setOnKeyListener { v, keyCode, event ->
-            if (delegateOnKeyListener?.onKey(v, keyCode, event) == true) {
-                return@setOnKeyListener true
-            }
-            if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
-                handleBackPress()
-            } else {
-                false
-            }
-        }
-    }
-
-    private fun handleBackPress(): Boolean {
-        return when (store.state.value.navigation) {
+    override fun handleOnBackPressed() {
+        when (store.state.value.navigation) {
             NavigationState.FreeDrive -> {
-                false
+                error("The callback should be disabled in Free Drive state")
             }
             NavigationState.DestinationPreview -> {
                 store.dispatch(DestinationAction.SetDestination(null))
                 store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
-                true
             }
             NavigationState.RoutePreview -> {
                 store.dispatch(RoutePreviewAction.Ready(emptyList()))
                 store.dispatch(NavigationStateAction.Update(NavigationState.DestinationPreview))
-                true
             }
             NavigationState.ActiveNavigation -> {
                 store.dispatch(RoutesAction.SetRoutes(emptyList()))
                 store.dispatch(NavigationStateAction.Update(NavigationState.RoutePreview))
-                true
             }
             NavigationState.Arrival -> {
                 store.dispatch(endNavigation())
-                true
             }
         }
     }
 
-    override fun onDetached(mapboxNavigation: MapboxNavigation) {
-        super.onDetached(mapboxNavigation)
-        view.isFocusableInTouchMode = false
-        view.setOnKeyListener(null)
+    companion object {
+
+        fun attach(
+            componentActivity: ComponentActivity,
+            store: Store,
+            lifecycleOwner: LifecycleOwner,
+        ) {
+            val onBackPressedCallback = DefaultOnBackPressedCallback(store)
+            componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
+            lifecycleOwner.lifecycleScope.launch {
+                try {
```

Comment:
Feels wrong to have try and finally when Flow is designed to control all these things
``` kotlin
lifecycleOwner.lifecycleScope.launch {
    lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
        store.select { it.navigation }
            .map { state -> state != NavigationState.FreeDrive }
            .distinctUntilChanged()
            .onCompletion { onBackPressedCallback.remove() }
            .collect { onBackPressedCallback.isEnabled = it }
    }
}
```

Reply:
With your suggestion the callback is removed in `inStop`, but I would to remove it in `onDestroy`.

Reply:
Yeah this was browser coding, point am making is the `try` `finally` and use the Flow operators.

Would this work?

```
lifecycleOwner.lifecycleScope.launch {
    lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
        val onBackPressedCallback = DefaultOnBackPressedCallback(store)
            componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
        store.select { it.navigation }
            .map { state -> state != NavigationState.FreeDrive }
            .distinctUntilChanged()
            .onCompletion { onBackPressedCallback.remove() }
            .collect { onBackPressedCallback.isEnabled = it }
    }
}
```

Reply:
Resolving cause this is just ideas